Add global inserter to Nav editor screen

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

Adds the global inserter to the Navigation block so as to achieve parity with other editors and also to provide users with an alternative means of adding new items.

**Question**: do we want/need e2e tests?

Addresses part of https://github.com/WordPress/gutenberg/issues/34268

Closes https://github.com/WordPress/gutenberg/issues/34438

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

> What I found jarring is that unlike any other block the navigation links automatically open that LinkUI which is totally unexpected when adding from the inserter (cc @javierarce)

Note that this problem is not a facet of this PR. Rather it's a general problem for the Nav Editor. It should be solved in another PR.

To test:

* Enable experiment and go to new Navigation screen - you should see the global inserter in the top toolbar. It may be disabled.
* Create an empty menu - global inserter should be disable until you choose an option from the placeholder.
* Click `Start blank`.
* Use the global inserter to insert some blocks - it should close each time a block is inserted. **Question**: we need this behaviour on smaller screens where the panel obscures the main content area _but_ do we want this on larger screens?
* Try adding nested blocks - check that using the global inserter adds blocks in the expected place.
* Try selecting the top nav block - check that using the global inserter adds blocks after the last nav item on the top level.
* Try to select nothing (not even the Nav block) - check that the global inserter is not disabled and using the global inserter adds blocks after the last nav item on the top level.
* Add the following code to create a Custom Post type - this is required to trigger the `Browse All` below.
```php
function myplugin_register_book_post_type() {
	$args = array(
		'public'       => true,
		'label'        => 'Books',
		'show_in_rest' => true,
	);
	register_post_type( 'book', $args );
}
add_action( 'init', 'myplugin_register_book_post_type' );
```
* Use the Quick Inserter and click `Browser All`. 
    - Check this triggers the Global Inserter. 
    - Check that when you choose a block it is inserted at the location you were on when you triggered via the Quick Inserter.
* Check the mobile experience by repeating the above.

## Screenshots <!-- if applicable -->


https://user-images.githubusercontent.com/444434/132684888-3b814f6a-0e32-4c6f-aab4-e1498ee42377.mov



## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Wire up store state and UI

getdave committed:
Add inserter button to header toolbar

getdave committed:
Use correct classname

Mamaduka commented:
This might make this a little confusing if/when we add #34227.

draganescu commented on packages/edit-navigation/src/components/layout/index.js:
```diff
@@ -43,8 +44,11 @@ const interfaceLabels = {
+	secondarySidebar: __( 'Block library' ),
 };
 
+const isInserterOpened = true;
```
why does this default to true?

getdave replied:
Oh that's probably an artifact from when I was hacking together the initial implementation. I had to port over a number of variables and I ended up hardcoding them whilst I worked out the correct state, actions and selectors to make the values dynamic.

draganescu commented:
What I found jarring is that unlike any other block the navigation links automatically open that LinkUI which is totally unexpected when adding from the inserter (cc @javierarce)

Also I noticed that if nothing is focused then the inserter is empty. I definitely remember we fixed this for widgets as well since it's a problem with having no post as final parent or something.

getdave committed:
Use optimal props on inserter toggle

getdave committed:
Allow global inserter to insert blocks at root level of Nav block

getdave committed:
Avoid displaying inserter if there are no blocks to insert

getdave committed:
Hide previews as they don't provide much value here

talldan commented on packages/edit-navigation/src/hooks/use-navigation-editor-insertion-point.js:
```diff
@@ -0,0 +1,21 @@
+		const { getBlockOrder } = select( blockEditorStore );
+
+		return {
+			rootClientId: getBlockOrder()[ 0 ],
```
I think this should only be the case when no block is selected or the nav block is selected, otherwise it should be possible to fall back to the block library default behavior of inserting after the selected block.

getdave commented:
> This might make this a little confusing if/when we add #34227.

Good spot. I assume you mean in terms of having x2 icons? I think @javierarce is going to create a dedicate text-based UI for "Create". TBH I think we _need_ to use text in that case because it's not part of the standard Gutenberg UI so we can't rely on icons.

getdave committed:
Refactor Inserter to dedicate component

getdave committed:
Remove hardcoded conditional

getdave committed:
Remove unwanted ref usage

getdave committed:
Add toggle button styles and remove close button on larger screens

Mamaduka commented:
@getdave I noticed that PR also contains changes for two `block-library` components. Maybe we should extract those into separate smaller PRs?

getdave committed:
Allow tools to show on mobile viewports

getdave committed:
Only hide undo/redo on smaller viewports

getdave committed:
Simplify Navigation Block hook

getdave committed:
Ensure global inserter always targets root nav block

getdave commented:
> @getdave I noticed that PR also contains changes for two `block-library` components. Maybe we should extract those into separate smaller PRs?

<img width="905" alt="Screen Shot 2021-09-08 at 13 57 11" src="https://user-images.githubusercontent.com/444434/132513355-56946322-bf9b-489f-a47c-2d21ab63279a.png">

Agreed 😄

getdave commented:
@Mamaduka Here is the extracted changes for block editor https://github.com/WordPress/gutenberg/pull/34656

spacedmonkey commented on packages/edit-navigation/src/components/secondary-sidebar/inserter-sidebar.js:
```diff
@@ -0,0 +1,73 @@
+							: undefined
+					}
+					showInserterHelpPanel={ SHOW_PREVIEWS }
+					showSearch={ false }
```
Why is search hidden here? Isn't the plan to have more core blocks, like social icons, search etc and third party blocks in this inserter? Meaning that search might be useful. Could be only hide the search, if there are more than 8 blocks or something?

talldan replied:
Yeah, I agree with @spacedmonkey's comment.

Also have to consider the variations that get created from taxonomies and custom post types which could inflate the available blocks quite a bit.

getdave replied:
Good question to ask!

I've hidden it mainly because as a default experience it seemed completely OTT to have a search. 

Thinking again:

* Your point about more blocks in the future is valid. I considered it but decided we could conditionally reintroduce it later if necessary. I might be overcomplicating things I think.
* Users of assistive tech _might_ still rely on search to find blocks.

Either way [I've extracted the changes requires to disable search in the block library to a separate PR](https://github.com/WordPress/gutenberg/pull/34656) and will remove those commits from this PR so we don't have to consider it now.

I still think the search is visually very dominant given that by default we have a limited number of blocks. But I don't feel super strongly about it.

getdave replied:
Update: I removed the changes from this PR that hide the search bar. If we need to deal with that UX then we can handle in a follow up.

javierarce commented:
> I think @javierarce is going to create a dedicate text-based UI for "Create".

Yes, here's the updated the design: https://github.com/WordPress/gutenberg/issues/31241#issuecomment-915871085

getdave committed:
Ensure items always inserted at end if root nav block is selected

getdave committed:
Update to use correct exported variable name

getdave committed:
Handle focus and tab trapping via hook

talldan commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -168,8 +168,8 @@ function Iframe( { contentRef, children, head, tabIndex = 0, ...props }, ref ) {
 	const [ bodyClasses, setBodyClasses ] = useState( [] );
-	const styles = useParsedAssets( window.__editorAssets.styles );
-	const scripts = useParsedAssets( window.__editorAssets.scripts );
+	const styles = useParsedAssets( window.__editorAssets?.styles );
+	const scripts = useParsedAssets( window.__editorAssets?.scripts );
```
How do these relate to the change in the PR?

getdave replied:
Ah no. Great spot. I forgot to remove those changes. Let me check if it's still causing problems.

getdave replied:
Changeset removed.

talldan commented on packages/edit-navigation/src/components/layout/index.js:
```diff
@@ -171,6 +184,7 @@ export default function Layout( { blockEditorSettings } ) {
 									<ComplementaryArea.Slot scope="core/edit-navigation" />
 								)
 							}
+							secondarySidebar={ renderSecondarySidebar() }
```
Could this be simplified?
```suggestion
							secondarySidebar={ isInserterOpened && ( <InserterSidebar /> ) }
```

getdave replied:
For now...yes. In the future we'll probably want to introduce List View so I'm following the pattern already established in other editors which is to have a dedicate function to determine the secondary sidebar.

I'll simplify for now.

getdave replied:
✅

talldan commented on packages/edit-navigation/src/components/header/inserter.js:
```diff
@@ -0,0 +1,55 @@
+		<ToolbarItem
+			ref={ inserterButton }
+			as={ Button }
+			className="edit-navigation-header-toolbar__inserter-toggle"
```
So the CSS naming guidelines are that the classname should represent the package and component name, only sub-elements within the component should have an `__element`. 

That makes sense because the component should be reusable (at least within the header), so may not always be rendered inside the 'header-toolbar' part of the UI.

```suggestion
			className="edit-navigation-inserter"
```

That name could be too generic though, so an idea is to give the component a prefixed name based on the folder:
- `HeaderInserterToggle`
- `edit-navigation-header-inserter-toggle`

I feel like the toggle suffix is also worth it, as my mistake was to initially think this component is the inserter itself.

getdave replied:
I'll rename the component as well.

spacedmonkey replied:
There were issues with having a different css class for the inserter, with widget block editor and e2e tests. Just want to make sure we can make e2e utils like insert block ( that uses search to find the block btw ) will work here as well.

getdave replied:
I've updated the component to be `InserterToggle` and updated the CSS to `edit-navigation-header-inserter-toggle`.

getdave replied:
I'll check the e2e tests don't break and we can still locate the inserter. If we're relying on classnames to locate the inserter we need to rethink how we're writing our tests. If that's the case I'll rewrite to make the tests less dependent on developer APIs.

talldan commented on packages/edit-navigation/src/components/secondary-sidebar/inserter-sidebar.js:
```diff
@@ -0,0 +1,89 @@
+/**
```
My personal preference would be for the path to this file to be `packages/edit-navigation/src/components/inserter-sidebar/index.js`, as I'd find that easier to locate.

Reasoning - Secondary sidebar is where the inserter sidebar is rendered, but similar to the classname issue, if it were ever rendered anywhere else the folder name would have to be changed.

getdave replied:
Good point. I was feeling a little uneasy about this. Your comment has made my concerns clearer.

getdave replied:
✅

talldan commented on packages/edit-navigation/src/components/secondary-sidebar/inserter-sidebar.js:
```diff
@@ -0,0 +1,89 @@
+		return null;
+	}
+
+	const isNavBlockSelected = navBlockClientId === selectedBlockClientId;
+
+	return (
+		<div
+			ref={ inserterDialogRef }
+			{ ...inserterDialogProps }
+			className="edit-navigation-layout__inserter-panel"
+		>
+			<div className="edit-navigation-layout__inserter-panel-header">
+				<Button
+					icon={ close }
+					onClick={ () => setIsInserterOpened( false ) }
+				/>
+			</div>
+			<div className="edit-navigation-layout__inserter-panel-content">
+				<Library
+					shouldFocusBlock={ isMobileViewport }
+					rootClientId={ navBlockClientId }
+					// If the root Nav block is selected then any items inserted by the
+					// global inserter should append after the last nav item. Otherwise
+					// simply allow default Gutenberg behaviour.
+					__experimentalInsertionIndex={
+						isNavBlockSelected ? lastNavBlockItemIndex : undefined
+					}
```
I think there are two cases where the end of the nav block should be inserted into. That's when the nav block is selected or when no blocks are selected. In both of those situations the default inserter behavior would be to insert after the nav block, which is not possible because that's a locked template.

Always providing `rootClientId` doesn't feel right to me because it's like we're telling `Library` to always insert in the nav block's root level, which is not the case when inserting into submenus. I'm a bit surprised it works, but I think it could be avoided for safety.

Something like this could make it clearer:
```jsx
const shouldInsertInNavBlock = ! selectedBlockClientId || navBlockClientId === selectedBlockClientId;
// ...
<Library
    shouldFocusBlock={ isMobileViewport }
    rootClientId={ shouldInsertInNavBlock ? navBlockClientId : undefined }
    __experimentalInsertionIndex={
        shouldInsertInNavBlock ? lastNavBlockItemIndex : undefined
    }
 ```

getdave replied:
Weirdly this is almost exactly what I had locally previously but I reverted it because I found it worked without it (ie: what you see in the PR now). I was also surprised it worked!

Let's be safe and do what you suggested.

getdave replied:
This is handy 

https://github.com/WordPress/gutenberg/blob/c5758d3fd125ffe14f00a25ba5b1c8080dfc33e4/packages/block-editor/src/components/inserter/hooks/use-insertion-point.js#L20-L32

getdave replied:
✅

talldan commented on packages/edit-navigation/src/hooks/use-navigation-editor-root-block.js:
```diff
@@ -0,0 +1,25 @@
+ * Internal dependencies
+ */
+
+const useNavigationEditorRootBlock = () => {
```
This is probably small enough to be inlined in the InserterSidebar component.

getdave replied:
I think it's used in x2 places though:

1. https://github.com/WordPress/gutenberg/pull/34619/files#diff-377c1b386160065e7a9fedf695210e9f1e8f2b62b0b0ccd304e6b886527738d1R19
2. https://github.com/WordPress/gutenberg/pull/34619/files#diff-a4e73f041700a24dfbe8f2a50b6dbaafa10b80022b807b4caf29db2df9e86d92R20

talldan replied:
Oh, good point, ignore this then!

talldan reviewed (commented):
Looks great so far, nice work!

Something I noticed is that when I registered an extra CPT to increase the inserter items to 7, the quick inserter was missing the 'Browse all' button that usually appears at the bottom.

Nav editor:
<img width="443" alt="Screenshot 2021-09-10 at 2 16 50 pm" src="https://user-images.githubusercontent.com/677833/132811008-201a7f48-ddd6-4bdc-8353-b845b697f361.png">

Post editor:
<img width="350" alt="Screenshot 2021-09-10 at 2 39 06 pm" src="https://user-images.githubusercontent.com/677833/132811022-99b715cb-ec55-4bc3-a99e-a2447d9fc57e.png">

Here the code snippet I used for the CPT (for some reason it results in two Post Link variations, but that's a separate issue):
```php
function myplugin_register_book_post_type() {
	$args = array(
		'public'       => true,
		'label'        => 'Books',
		'show_in_rest' => true,
	);
	register_post_type( 'book', $args );
}
add_action( 'init', 'myplugin_register_book_post_type' );
```

Would be great to look into getting the Browse all button working.

getdave committed:
Rename component to make it clear it's the inserter button

getdave committed:
Alter directory structure in order to improve portability of code

spacedmonkey commented on packages/edit-navigation/src/store/reducer.js:
```diff
@@ -98,8 +98,23 @@ export function selectedMenuId( state = null, action ) {
+ * @param {boolean|Object} state
+ * @param {Object}         action
+ */
+function blockInserterPanel( state = false, action ) {
```
Is it cleaner to deconstruct with object? 
```suggestion
function blockInserterPanel( state = false, { type, value } ) {
```

getdave replied:
For now, I'm following the pattern established elsewhere in the file.

spacedmonkey commented on packages/edit-navigation/src/store/reducer.js:
```diff
@@ -98,8 +98,23 @@ export function selectedMenuId( state = null, action ) {
+/**
+ * Reducer tracking whether the inserter is open.
+ *
+ * @param {boolean|Object} state
+ * @param {Object}         action
```
Could we add some comments here?

getdave replied:
✅

spacedmonkey commented on packages/edit-navigation/src/store/selectors.js:
```diff
@@ -81,3 +81,14 @@ export const getMenuItemForClientId = createRegistrySelector(
+ * @return {boolean} Whether the inserter is opened.
+ */
+export function isInserterOpened( state = false ) {
+	return !! state.blockInserterPanel;
```
Should this be conditional?
```suggestion
	return !! state?.blockInserterPanel;
```

getdave replied:
The reducer defines a default state value of `false` so I think we're ok here but there's no harm so I've updated.

getdave committed:
Make insertion point logic explicit

getdave committed:
Add docblocks to store

getdave committed:
Add selector state guard

spacedmonkey reviewed (approved):
👍

getdave committed:
Implement Browse All on quick inserter.

Required refactor of initial file.

getdave commented:
> Would be great to look into getting the Browse all button working.

@talldan I have the `Browse all` working for ya. 

https://user-images.githubusercontent.com/444434/132840628-a975bf93-4490-4b62-842c-b6d3e6837a32.mov

Required a refactor of the editor initialisation logic (to allow me to use the `useDispatch` hook) so I went ahead and refactored that a little to match other more complex editors.

Mamaduka commented on packages/edit-navigation/src/index.js:
```diff
@@ -5,8 +5,13 @@ import {
+/**
+ * Internal dependencies
+ */
+import { store as editNavigationStore } from './store';
```
I think Prettier messed up here. We have 3 "Internal dependencies" docblocks in this file.

getdave replied:
✅ 😆

Mamaduka commented on packages/edit-navigation/src/components/inserter-sidebar/index.js:
```diff
@@ -0,0 +1,95 @@
+import { useNavigationEditorRootBlock } from '../../hooks';
+
+function InserterSidebar() {
+	const SHOW_PREVIEWS = false;
```
Do we still need this constant? We can pass `false` directly into `showInserterHelpPanel`. But If you preferer having a constant, we should move it outside of the function. No need to recreate it on every rerender.

getdave replied:
I would like the constant purely as a way to make the code more self documenting. I'll move outside.

getdave replied:
✅

Mamaduka commented on packages/edit-navigation/src/components/inserter-sidebar/index.js:
```diff
@@ -0,0 +1,95 @@
+				).getSelectedBlock()?.clientId,
+			};
+		},
+		[]
```
```suggestion
		[ navBlockClientId ]
```

Missing dependency.

getdave replied:
✅

Mamaduka commented on packages/edit-navigation/src/components/header/inserter-button.js:
```diff
@@ -0,0 +1,55 @@
+			),
+			isInserterOpened: select( editNavigationStore ).isInserterOpened(),
+		};
+	} );
```
```suggestion
	}, [ navBlockClientId ] );
```

getdave replied:
✅

Mamaduka commented on packages/edit-navigation/src/components/header/inserter-button.js:
```diff
@@ -0,0 +1,55 @@
+
+	return (
+		<ToolbarItem
+			ref={ inserterButton }
```
This `ref` is probably a leftover from Edit Post screen implementation.

getdave replied:
Yeh I re-added it. Now removed it.

Mamaduka commented:
Thanks for working on this, Dave.

Inserter looks nice, and I like how it handles insertion points.

I left minor code-related comments.

I also noticed that screen jumps a lot when a new menu item is inserted. We can address this in follow-up PRs, but I wanted to mention it here.

https://user-images.githubusercontent.com/240569/133073450-d26e02e9-2a59-402e-98fe-f7c7ba1e38d1.mp4

getdave commented:
> I also noticed that screen jumps a lot when a new menu item is inserted. We can address this in follow-up PRs, but I wanted to mention it here.

Yeh that pesky link UI automatically appearing causes a lot of issues. Let's look to address in a followup. Let me tidy up those code fixes for you.

getdave committed:
Move constant outside component body

Addresses https://github.com/WordPress/gutenberg/pull/34619#discussion_r707217856

getdave committed:
Add dep to useSelect

getdave committed:
Add another dep to useSelect

getdave committed:
Remove ref

getdave committed:
Sort out Prettier's mess

getdave committed:
Scaffold out very basic interface

getdave committed:
Conditionalise render of sidebar

Ensures main content is visible on mobile devices when sidebar is disabled. If we return a component to the secondarySidebar then the interface component will render the wrapping div which will obscure the wrapping content.

getdave committed:
Simplify secondary sidebar

Mamaduka reviewed (approved):
@getdave, I think this looks good to merge.

We can address screen jumpiness after menu insertion in follow-up PRs.

getdave committed:
Rename Inserter Button to Inserter Toggle for clarity and consistency

getdave committed:
Apply correct CSS naming convention

Addresses https://github.com/WordPress/gutenberg/pull/34619#discussion_r705912360

getdave commented:
I'm going to raise a follow up here for e2e tests
